### ISSUE TITLE
`Optional` not rendered correctly as union in a return type

### ISSUE THREAD

**Issue by kmillikin** [microsoft/pyright#3307]
I noticed that this:

```python
def f() -> NoReturn: raise AttributeError()

reveal_type(cast(Union[Callable[[], None], Callable[[], Union[int, str]]], f))
```

renders as `() -> None | () -> (int | str)`, suggesting that `|` has lower priority than `->` (which is what I would expect, anyway).  But then this:

```python
reveal_type(cast(Union[Callable[[], int], Callable[[], Optional[str]]], f))
```

renders the `Optional[str]` return type incorrectly as a union: `() -> int | () -> str | None`.

Edit: I expect `() -> int | () -> (str | None)`.

**Comment by kmillikin** [comment]
This was observed in an old version of pyright.  Today at version 1.1.235 I observe

```
(() -> None) | (() -> (int | str))
```

and

```
(() -> int) | (() -> str | None)
```

which now seems to indicate that `->` has lower precedence than `|`, contains an unnecessary set of parentheses around `int | str` in the first case, and is inconsistent between the two cases.  But no longer ambiguous assuming the intended precedence of `->` and `|` is as I indicated here.

**Comment by erictraut** [comment]
Yes, I agree this is inconsistent. The "type printer" has some special-case logic to handle `None` because it used to be rendered as `Optional[x]`. This optional code path didn't add parentheses. I've fixed it so it's now consistent.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.236, which I just published. It will also be included in the next release of pylance.
